crates/transparency/src/map/proof.rs

```diff
@@ -28,72 +29,75 @@ use super::{
 ///
 /// Third, since sparse peers are more likely at the bottom of the tree, we
 /// can omit all leading sparse peers. The verifier can dynamically reconstruct
-pub struct Proof<D, V>
+pub struct Proof<D, K, V>
 where
     D: SupportedDigest,
+    K: VisitBytes,
     V: VisitBytes,
 {
+    key: PhantomData<K>,
     value: PhantomData<V>,
-    peers: Vec<Option<Hash<D>>>,
+    /// Sibling node hashes needed to construct a proof
+    pub peers: Vec<Option<Hash<D>>>,
 }
 
-impl<D, V> Proof<D, V>
+impl<D, K, V> Proof<D, K, V>
 where
     D: SupportedDigest,
+    K: VisitBytes,
     V: VisitBytes,
 {
     pub(crate) fn new(peers: Vec<Option<Hash<D>>>) -> Self {
         Self {
+            key: PhantomData,
             value: PhantomData,
             peers,
         }
     }
 
     pub(crate) fn push(&mut self, peer: Option<Hash<D>>) {
-        // This is an optimization. The size of a proof is always
-        // known: it is the number of bits in the digest. Therefore,
-        // we can skip all leading nodes with no peer. The validator,
-        // can reconstruct this.
-        if !self.peers.is_empty() || peer.is_some() {
-            self.peers.push(peer);
-        }
-    }
-
-    #[cfg(test)]
-    pub(crate) fn len(&self) -> usize {
-        self.peers.len()
+        self.peers.push(peer);
     }
 
     /// Computes the root obtained by evaluating this inclusion proof with the given leaf
-    pub fn evaluate<K: ?Sized + VisitBytes>(&self, key: &K, value: &V) -> Hash<D> {
+    pub fn evaluate(&self, key: K, value: &V) -> Hash<D> {
```

Comment:
We can still take `key` as `&K` if `ReversePath::new` takes `&K` (which it can).